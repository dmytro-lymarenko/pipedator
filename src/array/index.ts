import { test } from '../test';

export function array<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => Array.isArray(value), message || 'Value should be an array');
}
