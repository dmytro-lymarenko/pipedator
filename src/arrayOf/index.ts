import { every } from '../every';

/**
 * An alias for {@link every}
 */
export const arrayOf = every;
