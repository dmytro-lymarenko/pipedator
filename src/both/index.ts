/**
 * @module validators
 */

import { pipe } from '../pipe';

export const both = pipe;
