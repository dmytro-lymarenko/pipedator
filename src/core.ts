export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
