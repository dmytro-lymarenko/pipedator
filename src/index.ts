// This file is generated automatically by `scripts/buildIndex.ts`. Please, don't change it.

export * from './abstractShape';
export * from './alt';
export * from './alternative';
export * from './both';
export * from './core';
export * from './either';
export * from './empty';
export * from './equalTo';
export * from './every';
export * from './failure';
export * from './greater';
export * from './ignore';
export * from './keys';
export * from './nullable';
export * from './optional';
export * from './pipe';
export * from './shape';
export * from './some';
export * from './string';
export * from './success';
export * from './ternary';
export * from './test';
export * from './tuple';
export * from './valid';
export * from './values';
export * from './valuesByKeys';
export * from './when';
