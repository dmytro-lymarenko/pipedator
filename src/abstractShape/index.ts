import { createValidator, findFirstError, isValidationError, Validator } from '../core';

export function abstractShape<Key>(
	keys: Key[],
	shape: (key: Key) => Validator,
	options?: { onlyFirstError?: boolean },
	message?: string
) {
	return createValidator({
		validate: (value, ctx) => {
			if (options && options.onlyFirstError) {
				const { error } = findFirstError(i => shape(keys[i]), i => value[keys[i]], keys.length, i => ({
					...ctx,
					path: [...ctx.path, keys[i].toString()],
				}));

				if (error && message) {
					return ctx.generateError({
						value,
						message,
						path: ctx.path,
					});
				}

				return error;
			}

			const errors = keys
				.map(key =>
					shape(key).validate(value[key], {
						...ctx,
						path: [...ctx.path, key.toString()],
					})
				)
				.filter(isValidationError)
				.reduce((res, err) => [...res, ...err.errors], []);

			if (errors.length > 0) {
				if (message) {
					return ctx.generateError({
						value,
						message,
						path: ctx.path,
					});
				}

				return ctx.generateErrors(errors);
			}

			return null;
		},
	});
}
