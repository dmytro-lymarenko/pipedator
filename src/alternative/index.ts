/**
 * @module validators
 */

import { createValidator, getFirstErrors, validationErrorToString, Validator } from '../core';

export function alternative(validators: Validator[], message?: string) {
	return createValidator({
		validate: (value, ctx) => {
			if (validators.length === 0) {
				return null;
			}
			// find the first success validator
			const errors = getFirstErrors(i => validators[i], () => value, validators.length, () => ctx);

			if (errors.length === validators.length) {
				// here all validators returned an error
				return ctx.generateError({
					value,
					message:
						message ||
						`At least one validator should succeed: [${errors.map(error => validationErrorToString(error)).join(' OR ')}]`,
					path: ctx.path,
				});
			}

			return null;
		},
	});
}
