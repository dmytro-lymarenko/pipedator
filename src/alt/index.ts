import { alternative } from '../alternative';

export const alt = alternative;
