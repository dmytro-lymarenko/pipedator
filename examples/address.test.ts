import { shape, string, number, values, empty, ternary, optional } from '../src';

/*
This example dedonstrates how to build address validator with the next requirements:
  - it must be an object with a set of properties (email, country, ip)
  - at least one field is required
*/

function addressValidator() {
	const model = {
		email: string(),
		country: string(),
		ip: number(),
	};

	return ternary(values(empty()), shape(model), shape(model, { wrapValidators: optional }));
}

describe('Address test', () => {
	it('should work', () => {
		// Should show errors for each field
		expect(addressValidator().validate({})).toMatchSnapshot();
		// Should show errors for each field
		expect(addressValidator().validate({ email: null, country: null, ip: null })).toMatchSnapshot();
		// should be valid
		expect(addressValidator().validate({ country: 'USA' })).toMatchSnapshot();
		// should fail and show one error: country is invalid
		expect(addressValidator().validate({ country: 1 })).toMatchSnapshot();
		// should fail and show two errors: country and ip are invalid
		expect(addressValidator().validate({ country: 1, ip: 'wrong ip' })).toMatchSnapshot();
	});
});
