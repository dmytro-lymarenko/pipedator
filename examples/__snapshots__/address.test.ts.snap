// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address test should work 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a string",
      "path": "email",
    },
    Object {
      "children": null,
      "message": "Value should be a string",
      "path": "country",
    },
    Object {
      "children": null,
      "message": "Value should be a number",
      "path": "ip",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`Address test should work 2`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a string",
      "path": "email",
    },
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a string",
      "path": "country",
    },
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a number",
      "path": "ip",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`Address test should work 3`] = `null`;

exports[`Address test should work 4`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a string",
      "path": "country",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`Address test should work 5`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a string",
      "path": "country",
    },
    Object {
      "children": null,
      "message": "Either Value should equal to undefined OR Value should be a number",
      "path": "ip",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;
